mismi-s3/src/Mismi/S3/Commands.hs

```diff
@@ -447,6 +474,49 @@ multipartUploadWorker e mpu file a (o, c, i) =
         pure $! Right $! PartResponse i m
 
 
+uploadRecursiveWithMode :: WriteMode -> FilePath -> Address -> EitherT UploadError AWS ()
+uploadRecursiveWithMode m src (Address buck ky) = do
+  es <- tryIO $ getFileStatus src
+  case es of
+    Left _ -> left $ UploadSourceMissing src
+    Right st -> unless (isDirectory st) . left $ UploadSourceNotDirectory src
+  files <- liftIO $ listRecursivelyLocal src
+  let prefixLen = L.length (src </> "a") - 1
+      outputAddrs = fmap (\fp -> Address buck (ky // Key (T.pack $ L.drop prefixLen fp))) files
+  mapM_ (uncurry (uploadWithMode m)) $ L.zip files outputAddrs
+
+-- | Like `listRecursively` but for the local filesystem.
+listRecursivelyLocal :: MonadIO m => FilePath -> m [FilePath]
+listRecursivelyLocal topdir = do
+  entries <- liftIO $ listDirectory topdir
+  (dirs, files) <- liftIO $ partitionM isDir $ fmap (topdir </>) entries
```

Comment:
Uploading everything that's not a directory might have some weird consequences (e.g., block device or FIFO) - I'd probably filter `files` further with [isRegularFile](https://hackage.haskell.org/package/unix-2.7.2.2/docs/System-Posix-Files.html#v:isRegularFile) either here or at the call site. (You could print warnings about these a la `cp -r`, though I also think it'd be fine to just fail.)

Reply:
Yup, good idea.

Reply:
I updated the partion function to so that it find directories (which it recurses into) and regular files (which it uploads). Everything that isn't a directory or a regular files is silently dropped.